lib/puma/server.rb

```diff
@@ -807,19 +806,21 @@ def read_body(env, client, body, cl)
 
     # A fallback rack response if +@app+ raises as exception.
     #
-    def lowlevel_error(e, env)
+    def lowlevel_error(e, env, status=500)
       if handler = @options[:lowlevel_error_handler]
         if handler.arity == 1
           return handler.call(e)
-        else
+        elsif handler.arity == 2
           return handler.call(e, env)
+        else
+          return handler.call(e, env, status)
         end
       end
 
       if @leak_stack_on_error
```

Comment:
Maybe we should just get rid of this option and tell people to write their own lowlevel_handler that leaks the stack if they prefer. Next release is 5.0 so we can break APIs.

Reply:
What do you think?

Reply:
In the use cases I've used Puma, we always had `lowlevel_error_handler`s and this default didn't matter. It's not as friendly for out-of-the-box handling though, when somebody is new and trying to do local development. I'd vote to keep good defaults here.

I think the ideal way of doing this would be the default config for `lowlevel_error_handler` is:

```ruby
      if @options[:environment] == "development" || @options[:environment] == "test"
        [status, {}, ["Puma caught this error: #{e.message} (#{e.class})\n#{e.backtrace.join("\n")}"]]
      else
        [status, {}, ["An unhandled lowlevel error occurred. The application logs may have details.\n"]]
      end
```

That way your custom code is in the configuration, where it belongs, and you're not threading this extra stack context for a single bit of functionality.

I'm not sure the best way to do that though, so I'd push back a little and say it's a better change to make outside of this PR.